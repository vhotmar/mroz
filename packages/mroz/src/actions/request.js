import { createAction } from "redux-actions";

export const request = createAction(
  "@@mroz/REQUEST_REQUEST",
  (id, key, { isPolling = false } = {}) => ({
    id,
    key,
    isPolling
  })
);

export const success = createAction(
  "@@mroz/REQUEST_SUCCESS",
  (id, key, data) => ({
    id,
    key,
    data,
    time: Date.now()
  })
);

export const fail = createAction("@@mroz/REQUEST_FAIL", (id, key, error) => ({
  id,
  key,
  error
}));

export const cancel = createAction("@@mrot/REQUEST_CANCEL", id => ({
  id
}));
