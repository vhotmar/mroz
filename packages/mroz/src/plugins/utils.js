import { over, lensProp, concat, merge } from "ramda";

export const addMiddleware = (middleware = []) => plugin =>
  over(lensProp("middleware"), concat(middleware), plugin);

export const addReducers = (reducers = {}) => plugin =>
  over(lensProp("reducers"), merge(reducers), plugin);

export const addHooks = (hooks = []) => plugin =>
  over(lensProp("hooks"), concat(hooks), plugin);

export const addEnhancers = (enhancers = []) => plugin =>
  over(lensProp("enhancers"), concat(enhancers), plugin);
