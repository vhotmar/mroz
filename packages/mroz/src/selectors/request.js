export const requestByKey = (state, key) => state.request.requestsByKey[key];
