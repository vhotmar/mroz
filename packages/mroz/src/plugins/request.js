import R from "ramda";
import { combineReducers } from "redux";
import { handleActions } from "redux-actions";
import { ajax as rxAjax } from "rxjs/observable/dom/ajax";

import { addDependencies } from "./observable";
import { addReducers } from "./utils";

import * as actions from "../actions";

export const addHeaders = headers => properties =>
  R.over(R.lensProp("headers"), R.merge(properties.headers), headers);

export function ajax(properties) {
  return rxAjax(
    addHeaders({
      "Content-Type": "application/json",
      Accept: "application/json"
    })(properties)
  );
}

const reducer = selector =>
  handleActions(
    {
      [actions.request.request]: (state, action) => {
        const id = selector(action);
        const prev = state[id];

        return {
          ...state,
          [id]: {
            isFinished: prev.isFinished ? prev.isFinished : false,
            isPending: true,
            hasError: prev ? prev.hasError : false,
            error: prev ? prev.error : undefined,
            isPolling: action.payload.isPolling,
            queryCount: prev ? prev.queryCount + 1 : 1,
            lastUpdated: prev ? prev.lastUpdated : undefined
          }
        };
      },
      [actions.request.success]: (state, action) => {
        const id = selector(action);
        const prev = state[id];

        return {
          ...state,
          [id]: {
            ...prev,
            isFinished: true,
            isPending: false,
            hasError: false,
            error: undefined,
            lastUpdated: action.payload.time
          }
        };
      },
      [actions.request.fail]: (state, action) => {
        const id = selector(action);
        const prev = state[id];

        return {
          ...state,
          [id]: {
            ...prev,
            isFinished: true,
            isPending: false,
            hasError: true,
            error: action.payload.error
          }
        };
      },
      [actions.request.cancel]: (state, action) => {
        const id = selector(action);
        const prev = state[id];

        return {
          ...state,
          [id]: {
            ...prev,
            isFinished: true,
            isPending: false
          }
        };
      }
    },
    {}
  );

export default () =>
  R.compose(
    addDependencies({ ajax }),
    addReducers({
      request: combineReducers({
        requestsById: reducer(action => action.payload.id),
        requestsByKey: reducer(action => action.payload.key)
      })
    })
  );
