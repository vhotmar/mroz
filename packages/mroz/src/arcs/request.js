import R from "ramda";
import Backoff from "backo";
import actions from "./actions";
import { Observable } from "rxjs";
import { request as requestActions } from "../actions";
import { addWrapper, addFilter, addNormalize } from "../epics/async";
import {
  request as requestSelectors,
  async as asyncSelectors
} from "../selectors/";

const retryableStatusCodes = [0, 408, 429, 503, 504];
const defaultBackoffConfig = {
  minDuration: 300,
  maxDuration: 2000,
  maxAttempts: 5
};

export default function request(
  {
    polling,
    backoff: backoffConfig = defaultBackoffConfig,
    refetchTimeout
  } = {}
) {
  return R.pipe(
    // Dispatch request, success, fail actions which are going to be handled by
    // reducer
    actions(
      ({ action, id }) =>
        requestActions.request(id, action.meta.request.key, {
          isPolling: !!polling
        }),
      (data, { action, id }) =>
        requestActions.success(id, action.meta.request.key, data),
      (error, { action, id }) =>
        requestActions.fail(id, action.meta.request.key, error)
    ),
    // Filter requests
    addFilter((stream$, { store }) =>
      stream$.filter(action => {
        const state = store.getState();
        const meta = action.meta.request;

        const request = requestSelectors.requestByKey(state, meta.key);

        // Allow - first time request
        if (!request) return true;

        const requestState = asyncSelectors.state(state, request.id);

        // Allow - force request
        if (meta.force) {
          if (requestState.isPending) {
            // Cancel pending request
            store.dispatch(requestActions.cancel(request.id));
          }

          return true;
        }

        // Disable - pending request
        if (requestState.isPending) {
          return false;
        }

        // Allow - requests which should be refetched
        if (
          refetchTimeout &&
          Number(Date.now()) - Number(request.latestUpdate) > refetchTimeout
        ) {
          return true;
        }

        return false;
      })
    ),
    // Ensure requests are going to be executed, retried, polled and cancelled
    addWrapper((stream$, { id, action$, action, store }) => {
      const cancelStream = action$
        .ofType(String(requestActions.cancel))
        .filter(action => action.payload.id === id)
        .map(() => {
          throw new Error("cancel");
        });

      const pollingStream = polling
        ? Observable.interval(polling).filter(() => {
            const state = store.getState();
            const meta = action.meta.request;

            const request = requestSelectors.requestByKey(state, meta.key);

            // Ensure we are not requesting more than needed
            if (Date.now() - request.latestUpdate < polling) return false;

            return true;
          })
        : Observable.empty();

      return Observable.race(
        // Cancelation
        stream$
          // Handling errors
          .retryWhen(errors$ => {
            const backoff = new Backoff({
              min: backoffConfig.minDuration || 300,
              max: backoffConfig.maxDuration || 5000
            });

            return (
              errors$
                // Repeat request with backoff if this is not server error
                .scan((errorCount, error) => {
                  if (
                    !R.contains(error.status, retryableStatusCodes) ||
                    errorCount > (backoffConfig.maxAttempts || 5)
                  ) {
                    throw error;
                  }

                  return errorCount + 1;
                }, 0)
                .delayWhen(() => Observable.timer(backoff.duration()))
            );
          }),
        cancelStream
        // Ensure polling
      ).repeatWhen(pollingStream);
    }),
    // If everything went correctly, normalize the value
    addNormalize(data => data.response)
  );
}
