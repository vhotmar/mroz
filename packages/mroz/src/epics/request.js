import R from "ramda";
import isFunction from "lodash/isFunction";

import * as arcs from "../arcs";
import async, { addNormalize } from "./async";

export default (actions, parameters, normalize) =>
  async(
    actions.request,
    config =>
      config.dependencies.ajax(
        isFunction(parameters) ? parameters(config) : parameters
      ),
    R.compose(
      arcs.actions(actions.success, actions.fail),
      normalize ? addNormalize(normalize) : x => x,
      arcs.request()
    )
  );
