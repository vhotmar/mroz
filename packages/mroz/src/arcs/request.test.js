import request from "./request";
import { Observable } from "rxjs";
import { asyncTest, REQUEST_TYPE } from "../epics/async.test.utils";

const state = (r = {}, a = {}) => ({
  request: {
    byKey: r
  },
  async: {
    byId: a
  }
});

describe("actions arc", () => {
  const constantDate = new Date("2017-06-13T04:41:20");
  let _now;
  const t = {
    config: request({ refetchTimeout: 2000 }),
    getState: () => state(),
    work: () =>
      Observable.of({
        response: { data: "foo" }
      }),
    action: {
      type: REQUEST_TYPE,
      meta: {
        request: {
          key: "sample.request"
        }
      }
    }
  };

  beforeAll(() => {
    _now = Date.now;

    /*eslint no-global-assign:off*/
    Date.now = () => Number(constantDate);
  });

  afterAll(() => {
    Date.now = _now;
  });

  it("should dispatch actions", () => {
    asyncTest({
      ...t,
      callback: res => {
        expect(res).toMatchSnapshot();
      }
    }).exec();

    asyncTest({
      ...t,
      work: () => Observable.throw("error"),
      callback: res => {
        expect(res).toMatchSnapshot();
      }
    }).exec();
  });

  it("should filter out pending request", () => {
    asyncTest({
      ...t,
      getState: () =>
        state(
          {
            "sample.request": {
              id: "id1"
            }
          },
          {
            id1: {
              isPending: true
            }
          }
        ),
      callback: res => {
        expect(res).toMatchSnapshot();
      }
    }).exec();
  });

  it("should force request", () => {
    const dispatchFunction = jest.fn();

    asyncTest({
      ...t,
      action: {
        type: REQUEST_TYPE,
        meta: {
          request: {
            key: "sample.request",
            force: true
          }
        }
      },
      getState: () =>
        state(
          {
            "sample.request": {
              id: "1234"
            }
          },
          {
            "1234": {
              isPending: true
            }
          }
        ),
      dispatch: dispatchFunction,
      callback: res => {
        expect(res).toMatchSnapshot();
        expect(dispatchFunction).toHaveBeenCalledTimes(1);
        expect(dispatchFunction).toHaveBeenCalledWith({
          payload: { id: "1234" },
          type: "@@mrot/REQUEST_CANCEL"
        });
      }
    }).exec();
  });

  it("should refetch request if passed refetchTimeout", () => {
    asyncTest({
      ...t,
      getState: () =>
        state(
          {
            "sample.request": {
              id: "id1",
              latestUpdate: Number(constantDate) - 3000
            }
          },
          {
            id1: {}
          }
        ),
      callback: res => {
        expect(res).toMatchSnapshot();
      }
    }).exec();
  });

  it("should not refetch request if refetchTimeout have not passed", () => {
    asyncTest({
      ...t,
      getState: () =>
        state(
          {
            "sample.request": {
              id: "id1",
              latestUpdate: Number(constantDate) - 1000
            }
          },
          {
            id1: {}
          }
        ),
      callback: res => {
        expect(res).toMatchSnapshot();
      }
    }).exec();
  });
});
