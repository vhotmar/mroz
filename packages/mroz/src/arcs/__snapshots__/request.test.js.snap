// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions arc should dispatch actions 1`] = `
Array [
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
    },
    "payload": Object {
      "id": "id",
      "isPolling": false,
      "key": "sample.request",
    },
    "type": "@@mroz/REQUEST_REQUEST",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
    },
    "payload": Object {
      "id": "id",
    },
    "type": "@@mroz/ASYNC_REQUEST",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "success",
      },
    },
    "payload": Object {
      "data": Object {
        "data": "foo",
      },
      "id": "id",
      "key": "sample.request",
      "time": 2017-06-13T04:41:20.000Z,
    },
    "type": "@@mroz/REQUEST_SUCCESS",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "success",
      },
    },
    "payload": Object {
      "data": Object {
        "data": "foo",
      },
      "id": "id",
    },
    "type": "@@mroz/ASYNC_SUCCESS",
  },
]
`;

exports[`actions arc should dispatch actions 2`] = `
Array [
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
    },
    "payload": Object {
      "id": "id",
      "isPolling": false,
      "key": "sample.request",
    },
    "type": "@@mroz/REQUEST_REQUEST",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
    },
    "payload": Object {
      "id": "id",
    },
    "type": "@@mroz/ASYNC_REQUEST",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "fail",
      },
    },
    "payload": Object {
      "error": "error",
      "id": "id",
      "key": "sample.request",
    },
    "type": "@@mroz/REQUEST_FAIL",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "fail",
      },
    },
    "payload": Object {
      "error": "error",
      "id": "id",
    },
    "type": "@@mroz/ASYNC_FAIL",
  },
]
`;

exports[`actions arc should filter out pending request 1`] = `Array []`;

exports[`actions arc should force request 1`] = `
Array [
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
    },
    "payload": Object {
      "id": "id",
      "isPolling": false,
      "key": "sample.request",
    },
    "type": "@@mroz/REQUEST_REQUEST",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
    },
    "payload": Object {
      "id": "id",
    },
    "type": "@@mroz/ASYNC_REQUEST",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "success",
      },
    },
    "payload": Object {
      "data": Object {
        "data": "foo",
      },
      "id": "id",
      "key": "sample.request",
      "time": 2017-06-13T04:41:20.000Z,
    },
    "type": "@@mroz/REQUEST_SUCCESS",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "success",
      },
    },
    "payload": Object {
      "data": Object {
        "data": "foo",
      },
      "id": "id",
    },
    "type": "@@mroz/ASYNC_SUCCESS",
  },
]
`;

exports[`actions arc should not refetch request if refetchTimeout have not passed 1`] = `Array []`;

exports[`actions arc should refetch request if passed refetchTimeout 1`] = `
Array [
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
    },
    "payload": Object {
      "id": "id",
      "isPolling": false,
      "key": "sample.request",
    },
    "type": "@@mroz/REQUEST_REQUEST",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
    },
    "payload": Object {
      "id": "id",
    },
    "type": "@@mroz/ASYNC_REQUEST",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "success",
      },
    },
    "payload": Object {
      "data": Object {
        "data": "foo",
      },
      "id": "id",
      "key": "sample.request",
      "time": 2017-06-13T04:41:20.000Z,
    },
    "type": "@@mroz/REQUEST_SUCCESS",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "success",
      },
    },
    "payload": Object {
      "data": Object {
        "data": "foo",
      },
      "id": "id",
    },
    "type": "@@mroz/ASYNC_SUCCESS",
  },
]
`;
